master/buildbot/util/githubapiservice.py

```diff
@@ -0,0 +1,231 @@
+                "exp": token_expire,
+                "iss": self.integration_id
+            }
+            self._jwt_token_expire = token_expire
+            if self.debug:
+                log.info('Generated JWT Token at {now}: {token}',
+                         now=datetime.datetime.utcnow(), token=payload)
+            self._jwt_token = jwt.encode(
+                payload,
+                key=self.private_key,
+                algorithm="RS256"
+            )
+        return self._jwt_token
+
+    @defer.inlineCallbacks
+    def get_access_token(self, user_id=None):
+        """
+        Get an access token for the given installation id.
+        POSTs https://api.github.com/installations/<installation_id>/access_tokens
+        """
+        if self._access_tokens.get(user_id, None) is not None:
+            expire_date = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
+            if self._access_tokens[user_id].expires_at < expire_date:
+                self._access_tokens.pop(user_id)
+        if self._access_tokens.get(user_id, None) is None:
+            body = None
+            if user_id:
+                body = {"user_id": user_id}
+
+            response = yield self.post(
+                "/installations/{}/access_tokens".format(self.installation_id),
+                headers={
+                    "Authorization": "Bearer {}".format(self.get_jwt_token()),
+                    "Accept": "application/vnd.github.machine-man-preview+json",
+                },
+                json=body
+            )
+
+            if response.code == 201:
+                data = yield response.json()
+                token = data['token']
+                expires_at = data['expires_at']
+                if len(expires_at) == 25:
+                    expires_at = datetime.datetime.strptime(
+                        expires_at[:19], "%Y-%m-%dT%H:%M:%S") + (
+                                1 if expires_at[19] == '-' else -1) * datetime.timedelta(
+                                        hours=int(expires_at[20:22]),
+                                        minutes=int(expires_at[23:25]))
+                else:
+                    expires_at = datetime.datetime.strptime(expires_at, "%Y-%m-%dT%H:%M:%SZ")
+                on_behalf_of = data.get('on_behalf_of', None)
+                self._access_tokens[user_id] = self.InstallationAuthorization(token, expires_at, on_behalf_of)
+            else:
+                body = yield response.content()
+                log.warn(
+                    'Unable to get GitHub installation auth token. HTTP Code: {code}. '
+                    'Error: {error}',
+                    code=response.code,
+                    error=body
+                )
+        if self._access_tokens.get(user_id, None) is not None:
+            defer.returnValue(self._access_tokens[user_id].token)
+        else:
+            defer.returnValue(None)
+
+    @defer.inlineCallbacks
+    def get(self, endpoint, **kwargs):
```

Comment:
You can see the httpclientservice implementation, which avoids to write 5 times the same thing to implement get, post, put , etc

Reply:
It was copy/paste from the real service.